File: src/main/java/org/cubexmc/metro/train/TrainTaskStarter.java
package org.cubexmc.metro.train;

import org.bukkit.Bukkit;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.model.Line;

/**
 * Creates and registers train movement tasks after a player boards.
 */
final class TrainTaskStarter {

    private TrainTaskStarter() {
    }

    static void start(Metro plugin, Minecart minecart, Player passenger, String lineId, String currentStopId) {
        LineManager lineManager = plugin.getLineManager();
        Line line = lineManager.getLine(lineId);
        if (line == null) {
            return;
        }

        if (passenger == null || !passenger.isOnline() || passenger.getVehicle() != minecart) {
            if (minecart.isValid()) {
                minecart.remove();
            }
            return;
        }

        TrainMovementTask trainTask = new TrainMovementTask(plugin, minecart, passenger, lineId, currentStopId);
        Bukkit.getPluginManager().registerEvents(trainTask, plugin);
        TrainTaskRegistry.register(minecart, trainTask);

        minecart.setMaxSpeed(0);
        minecart.setVelocity(new Vector(0, 0, 0));
        trainTask.startAtStation();
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainTaskStarter.kt
package org.cubexmc.metro.train

import org.bukkit.Bukkit
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.util.Vector
import org.cubexmc.metro.Metro

/**
 * Creates and registers train movement tasks after a player boards.
 */
object TrainTaskStarter {
    @JvmStatic
    fun start(plugin: Metro, minecart: Minecart, passenger: Player?, lineId: String, currentStopId: String) {
        val lineManager = plugin.lineManager
        val line = lineManager.getLine(lineId) ?: return

        if (passenger == null || !passenger.isOnline || passenger.vehicle != minecart) {
            if (minecart.isValid) {
                minecart.remove()
            }
            return
        }

        val trainTask = TrainMovementTask(plugin, minecart, passenger, lineId, currentStopId)
        Bukkit.getPluginManager().registerEvents(trainTask, plugin)
        TrainTaskRegistry.register(minecart, trainTask)

        minecart.maxSpeed = 0.0
        minecart.velocity = Vector(0, 0, 0)
        trainTask.startAtStation()
    }
}
